File: 2025-10-October-GFG-POTD/October 25 - Minimum Steps to Halve Sum/Solution.py
import heapq

class Solution:
    def minOperations(self, arr):
        original_sum = sum(arr)
        target = original_sum / 2.0
        current_sum = original_sum

        # Convert to max heap by storing negative values
        pq = [-float(x) for x in arr]
        heapq.heapify(pq)

        operations = 0

        while current_sum > target:
            largest = -heapq.heappop(pq)
            half = largest / 2.0
            current_sum -= half
            heapq.heappush(pq, -half)
            operations += 1

        return operations

File: 2025-10-October-GFG-POTD/October 25 - Minimum Steps to Halve Sum/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int minOperations(vector<int>& arr) {
        double originalSum = 0;
        priority_queue<double> pq;  // max-heap

        for (int num : arr) {
            originalSum += num;
            pq.push((double)num);
        }

        double target = originalSum / 2.0;
        double currentSum = originalSum;
        int operations = 0;

        while (currentSum > target) {
            double largest = pq.top();
            pq.pop();
            double half = largest / 2.0;
            currentSum -= half;
            pq.push(half);
            operations++;
        }

        return operations;
    }
};
